torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -0,0 +1,917 @@
+    getMergeCandidates() {
+  // Don't look for candidates if already merged
+  if (merged) {
+    return {};
+  }
+
+  std::vector<NeighborGroup> neighbors = getNeighborGroups();
+
+  // Can this node be merged with another? Check if neighbors are merged, if
+  // so and merged neighbor is within 1 level or node merged with neighbor is
+  // within 1 level, can't merge this node with anything else.
+  bool can_merge_this = true;
+  for (auto& neighbor : neighbors) {
+    if (!neighbor.group->merged) {
+      continue;
+    }
+    if (std::abs(neighbor.group->level - level) <= 1) {
+      can_merge_this = false;
+    }
+    if (std::abs(neighbor.group->merge_with->level - level) <= 1) {
+      can_merge_this = false;
+    }
+  }
+  if (!can_merge_this) {
+    return {};
+  }
+
+  std::vector<bool> can_merge(true, neighbors.size());
+
+  // Find neighbors with a level that is only 1 differant than this groups level
+  for (size_t i = 0; i < neighbors.size(); i++) {
+    if (std::abs(neighbors[i].group->level - level) > 1) {
+      can_merge[i] = false;
+    }
+  }
+
+  // Check neighbor of neighbors we're considering, if any of them are merged
+  // with another node, make sure the resulting edge wouldn't have a level
+  // difference of 1
+  for (size_t i = 0; i < neighbors.size(); i++) {
+    if (!can_merge[i]) {
+      continue;
+    }
+
+    for (auto neighbor_neighbor : neighbors[i].group->getNeighbors()) {
+      // Don't check self
+      if (neighbor_neighbor == neighbors[i].group) {
+        continue;
+      }
+      if (neighbor_neighbor->merged) {
+        // check neighbor_neighbor level
+        if (std::abs(neighbor_neighbor->level - level) <= 1) {
+          can_merge[i] = false;
+        }
+        if (std::abs(neighbor_neighbor->level - neighbors[i].group->level) <=
+            1) {
+          can_merge[i] = false;
+        }
+
+        // check neighbor_neighber->merged->level
+        if (std::abs(neighbor_neighbor->merge_with->level - level) <= 1) {
+          can_merge[i] = false;
+        }
+        if (std::abs(
+                neighbor_neighbor->merge_with->level -
+                neighbors[i].group->level) <= 1) {
+          can_merge[i] = false;
+        }
+      }
+    }
+  }
+
+  std::vector<NeighborGroup> merge_candidates;
+  for (size_t i = 0; i < neighbors.size(); i++) {
+    if (can_merge[i]) {
+      merge_candidates.push_back(neighbors[i]);
+    }
+  }
+  return merge_candidates;
+}
+
+void SegmentedGroup::clearTraversalInfo() {
+  level = -1;
+  visited = false;
+  merge_with = nullptr;
+  merged = false;
+}
+
+std::vector<Val*> SegmentedGroup::edgesToVals(
+    const std::vector<SegmentedEdge*>& se_v) {
+  std::vector<Val*> ret_v;
+  ret_v.reserve(se_v.size());
+
+  std::transform(
+      se_v.cbegin(),
+      se_v.cend(),
+      std::back_inserter(ret_v),
+      [](SegmentedEdge* se) { return se->val; });
+  return ret_v;
+}
+
+void SegmentedGroup::print() {
+  std::cout << "g{"
```

Comment:
this seems to duplicate the operator<< below - why not use that one here?

Reply:
Updated, Thanks!